### PR TITLE
run shellcheck as part of the acceptance test

### LINKED ISSUE #35: Adopt Shellcheck
As [mhurron mentions](https://www.reddit.com/r/sysadmin/comments/4pc4wu/bash3_boilerplate_template_for_writing_better/), it might be a good idea to adopt / advise on linting

### PR DESCRIPTION
**Pull request by mstreuhofer**
Please do not merge! This is a test to see Travis working (or not) - will update the PR once it is ready for merging.

---

shellcheck is only run if it is already installed.

right now only main.sh is fine. almost all other would fail which is why
the script acceptance.sh tests main.sh only.

this commit/PR is just to see if the whole things works in Travis as-is.

Closes #35

### CODE REVIEW

**Comment by mstreuhofer** [comment]
alright. that's what i thought. `shellcheck` is not installed in the Travis container and installing it (at least on OSX) takes _quite some time_. The Linux test could do it though.

**Comment by zbeekman** [comment]
> alright. that's what i thought. shellcheck is not installed in the Travis container and installing it (at least on OSX) takes quite some time. The Linux test could do it though.

Yes, I've struggled with shellcheck in the past because of the dependency on haskell etc... can do some research into how to get this up and running.... You may be able to `brew install --force-bottle` for shellckeck (and perhaps it's dependencies too....) on OS X. IIRC Linux may actually be the larger impediment. Some Travis-CI documentation claims it's whitelisted etc, but i thought that I remembered this to be a lie....

**Comment by zbeekman** [comment]
FYI the sudo disabled linux container is much faster in general, but perhaps switching to `sudo: required`, while slower, will be worth it for the shell check issue.... (Full disclosure, I have not looked at your commits yet)

**Comment by zbeekman** [comment]
if you get frustrated, I've done this in the past, so I may be able to do it more easily than you when I find time.

**Comment by mstreuhofer** [comment]
no worries. if this last test of mine now doesn't work i'll give up - for today. ;)
found your comments in the travis bug tracker. doesn't look like that anything changed until now.

**Comment by mstreuhofer** [comment]
YES 🎉 --> https://travis-ci.org/kvz/bash3boilerplate/jobs/182753700#L1482-L1483

**Comment by kvz** [comment]
I work with unfinished PRs a lot and find it a good way to encourage transparency and discussion early on. As a good practice, I would like to propose prefixing PR titles with `WIP` to indicate they are a work in progress, then removing that and adding a `Ready for review` comment when it's in good shape.

Regardless, I'm very much looking forward to adopting shellcheck in our test-suite, can't thank you enough for the work you put into this @mstreuhofer. If only you were on twitter I could publicly applaud you for it 😄 💯

File: test/acceptance.sh [hunk 1]
```diff
@@ -184,3 +186,34 @@ for scenario in $(echo ${scenarios}); do
...
+
+# finally do some shellcheck linting
+if [[ "$(command -v shellcheck)" ]]; then
+  echo "==> Shellcheck"
+  pushd "${__root}" > /dev/null
+
+  failed="false"
+
+  while IFS=$'\n' read -r file; do
+    [[ "$file" != "./main.sh" ]] && continue #XXX
```
Comment (zbeekman) [inline test/acceptance.sh]: just saw your top level comment.... yeah a bunch of stuff will need to be cleaned up... Also, I'm guessing shellcheck on OS X will be a no go... last time I tried installing with homebrew I ended up having to build from source...

**Comment by mstreuhofer** [comment]
so right now we know shellcheck is working in the linux container by installing it from debian sid repo. OSX most probably will be a no-go because the only thing i found about `--force-bottle` is that it is considered developer-only. and compiling it from source every time a test runs is a no-go because it takes way too long.

**Comment by zbeekman** [comment]
> OSX most probably will be a no-go because the only thing i found about --force-bottle is that it is considered developer-only. and compiling it from source every time a test runs is a no-go because it takes way too long.

Have you tried a simple `brew install shellckeck`? I actually just tried on my sierra system and it did **not** need to install `ghc` and it **did** pour from a bottle. Also, I use `--force-bottle` all the time... it's fine so long as either the package is relocatable or the homebrew layout matches the defaults.

**Comment by mstreuhofer** [comment]
@zbeekman i ran that brew install on my mavericks system and it took i believe like 90 minutes to compile it all. did no other checks beside that.

**Comment by zbeekman** [comment]
Yeah I mean there's probably no real reason to add it to OS X... it should
be doing the same thing on both systems...

Does homebrew still ship bottles for Mavericks? It may be too old

**Comment by mstreuhofer** [comment]
so i added code to `acceptance.sh` to run shellcheck. it does so (as of now) only for files which are whitelisted. right now this is `main.sh`, `example.sh` and `test/acceptance.sh`. See: https://github.com/kvz/bash3boilerplate/pull/77/files#diff-1c6fadfa430dff2933bd816025149891R208

Ideally that whitelisting should not exist. With time all other `*.sh` files can be fixed and the acceptance test extended. At the end the whole whitelist can be removed. I do not believe that all of this should be done in this PR here all at once.

So i guess this means this PR here is ready for review and merging.

File: test/acceptance.sh [hunk 2]
```diff
@@ -183,4 +189,41 @@ for scenario in $(echo ${scenarios}); do
...
+# finally do some shellcheck linting
+if [[ "$(command -v shellcheck)" ]]; then
+  echo "==> Shellcheck"
+  pushd "${__root}" > /dev/null
+
+  failed="false"
+
+  while IFS=$'\n' read -r file; do
+    lint="false"
+    [[ "$file" = "./main.sh" ]] && lint="true"
```
Comment (kvz) [inline test/acceptance.sh]: Missing braces here

File: test/acceptance.sh [hunk 3]
```diff
@@ -183,4 +189,41 @@ for scenario in $(echo ${scenarios}); do
...
+    [[ "$file" = "./example.sh" ]] && lint="true"
+    [[ "$file" = "./test/acceptance.sh" ]] && lint="true"
+    [[ "$lint" != "true" ]] && continue
+
+    echo -n "    ${file}.. "
+
+    if ! shellcheck --shell=bash --external-sources "$file" >> "${__accptstTmpDir}/shellcheck.err"; then
```
Comment (kvz) [inline test/acceptance.sh]: and here

Comment (mstreuhofer) [inline test/acceptance.sh]: you're right. will fix.
would be cool to our own set of rules in shellcheck too.
don't know if this is possible. will research.

**Pull request merged**